### DISCUSSION TITLE
Intercepting the audio samples

### DISCUSSION THREAD

**Discussion by KernalCrash** [schreibfaul1/ESP32-audioI2S#684] [Ideas]
Hi!

I just browsed all the readme, WiKi and discussions, but couldn't find how it would be possible to recycle the (for example) simple webradio sketch into feeding an FFT (arduinoFFT.h) to get a spectrum analyzer display driven by the audio data from a web radio stream.

In this case I don't need the I2S (because this interface gets used to hook up the LED matrix panel), just the converted into a mono audio stream data samples. Kinda so (half pseudo code):

```
#include <Audio.h>
...
Audio audio();
...
void setup () {
  ...
  audio.setVolume(10);
  audio.connecttohost("<streamingURL>"); // aac
}
...
void loop() {
  audio.loop();
  ...
  // Feed audio samples from web-stream instead of sampling the audio pin
  for (int i = 0; i < SAMPLES; i++) {
    newTime = micros();
    vReal[i] =   audio.getAudioSamples(leftChannel);  // analogRead(AUDIO_IN_PIN); // A conversion takes about 9.7uS on an ESP32
    vImag[i] = 0;
    while ((micros() - newTime) < sampling_period_us) { /* chill */ }
  }
}
```

Is there already a method or a possibility to do that or would that need yet to be implemented?

Thanks!

**Comment by KernalCrash** [comment]
... just to add, my most favorite candidate was

`
extern __attribute__((weak)) void audio_process_extern(int16_t* buff, uint16_t len, bool *continueI2S); // record audiodata or send via BT
`

but I just can't seem to wrap my head around as to how I could make use of that. Maybe I could then even use the second ESP32's I2S-I/F to still be able to playback the audio via an external DAC that I have.

**Comment by schreibfaul1** [comment]
This is exactly what you can use for this. This is the raw data from the decoder. You don't have to convert anything D/A. With mono, you just take every second value.
 ````c++ 
void audio_process_extern(int16_t* buff, uint16_t len, bool *continueI2S){
   for(int i = 0; i < len; i += 2){
         int16_t data = buff[i];
        // do soomething with data.... }
}
 ````

**Comment by KernalCrash** [comment]
So audio_process_extern gets called by the decoder and i just have to implement this function?

**Reply by KernalCrash** [reply]
Yes, that was it, works just fine now.
